File: ios/Sources/ForgeIOSRuntime/DerivedDataSourceRuntime.swift
import Foundation

public enum DerivedDataSourceRuntime {
    public static func run(
        sources: [String: [[String: JSONValue]]],
        spec: DerivedDataSourceSpec
    ) throws -> [[String: JSONValue]] {
        let limit = max(spec.maxRows ?? 10_000, 1)
        for ref in spec.sources where (sources[ref]?.count ?? 0) > limit {
            throw DerivedDataSourceError.message("Derived datasource source \(ref) exceeds maxRows \(limit); use a server datasource.")
        }
        var rows = sources[spec.sources.first ?? ""] ?? []
        for step in spec.pipeline {
            let operation = step.operation.lowercased()
            guard ["filter", "select", "map", "union", "join", "group", "sort"].contains(operation) else {
                throw DerivedDataSourceError.message("Unsupported derived operation: \(step.operation)")
            }
            if operation == "filter" {
                rows = rows.filter { row in
                    DashboardRuntime.evaluateDashboardCondition(step.when, form: row.compactMapValues(\.anyValue))
                }
            } else if operation == "select" || operation == "map" {
                rows = rows.map { deepMergeDerived($0, projectDerived(row: $0, step: step)) }
            } else if operation == "union" {
                let ref = try declaredDerivedSource(step.source, operation: operation, spec: spec)
                rows.append(contentsOf: sources[ref] ?? [])
            } else if operation == "join" {
                let ref = try declaredDerivedSource(step.source, operation: operation, spec: spec)
                rows = try joinDerived(left: rows, right: sources[ref] ?? [], step: step, limit: limit)
            } else if operation == "group" {
                rows = try groupDerived(rows: rows, step: step)
            } else if operation == "sort" {
                rows.sort { left, right in compareDerived(left, right, orderBy: step.orderBy ?? []) < 0 }
            }
            guard rows.count <= limit else {
                throw DerivedDataSourceError.message("Derived datasource output exceeds maxRows \(limit); use a server datasource.")
            }
        }
        return rows
    }

    private static func declaredDerivedSource(_ raw: String?, operation: String, spec: DerivedDataSourceSpec) throws -> String {
        let ref = raw?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !ref.isEmpty, spec.sources.contains(ref) else {
            throw DerivedDataSourceError.message("Derived \(operation) source \(ref.isEmpty ? "<empty>" : ref) must be declared in sources.")
        }
        return ref
    }

    private static func joinDerived(left: [[String: JSONValue]], right: [[String: JSONValue]], step: DerivedDataStepSpec, limit: Int) throws -> [[String: JSONValue]] {
        let fields = step.on?.isEmpty == false ? step.on! : ["id"]
        var rightByKey: [String: [[String: JSONValue]]] = [:]
        for row in right {
            let key = derivedIdentity(row, fields: fields)
            if step.joinCardinality?.lowercased() != "many", rightByKey[key]?.isEmpty == false {
                throw DerivedDataSourceError.message("Derived join expected one right row for key \(key).")
            }
            rightByKey[key, default: []].append(row)
        }
        var result: [[String: JSONValue]] = []
        for row in left {
            let matches = rightByKey[derivedIdentity(row, fields: fields)] ?? []
            if matches.isEmpty && step.joinType?.lowercased() == "inner" { continue }
            let selected = step.joinCardinality?.lowercased() == "many" ? (matches.isEmpty ? [[:]] : matches) : [matches.first ?? [:]]
            for match in selected {
                result.append(deepMergeDerived(row, projectDerived(row: match, step: step)))
                if result.count > limit { throw DerivedDataSourceError.message("Derived datasource output exceeds maxRows \(limit); use a server datasource.") }
            }
        }
        return result
    }

    private static func groupDerived(rows: [[String: JSONValue]], step: DerivedDataStepSpec) throws -> [[String: JSONValue]] {
        let groupFields = step.groupBy ?? []
        let grouped = Dictionary(grouping: rows) { derivedIdentity($0, fields: groupFields) }
        return try grouped.keys.sorted().map { key in
            let members = grouped[key] ?? []
            var result: [String: JSONValue] = [:]
            for field in groupFields { result[field] = resolveDerived(members.first ?? [:], field) ?? .null }
            for measure in step.measures ?? [] {
                let values = members.compactMap { row in measure.source.flatMap { resolveDerived(row, $0) } }.filter { $0 != .null }
                switch measure.operation.lowercased() {
                case "count": result[measure.target] = .number(Double(measure.source == nil ? members.count : values.count))
                case "sum": result[measure.target] = .number(values.compactMap(derivedNumber).reduce(0, +))
                case "min": result[measure.target] = values.min(by: { derivedComparable($0) < derivedComparable($1) }) ?? .null
                case "max": result[measure.target] = values.max(by: { derivedComparable($0) < derivedComparable($1) }) ?? .null
                case "first": result[measure.target] = values.first ?? .null
                case "list": result[measure.target] = .array(values)
                default: throw DerivedDataSourceError.message("Unsupported derived measure: \(measure.operation)")
                }
            }
            return result
        }
    }

    private static func projectDerived(row: [String: JSONValue], step: DerivedDataStepSpec) -> [String: JSONValue] {
        var result: [String: JSONValue] = [:]
        for (target, source) in step.fields ?? [:] {
            if case .string(let selector) = source { result[target] = resolveDerived(row, selector) ?? .null }
            else { result[target] = source }
        }
        for projection in step.projections ?? [] {
            result[projection.target] = projection.source.flatMap { resolveDerived(row, $0) } ?? projection.value ?? .null
        }
        return result
    }

    private static func compareDerived(_ left: [String: JSONValue], _ right: [String: JSONValue], orderBy: [JSONValue]) -> Int {
        for item in orderBy {
            guard let object = item.objectValue else { continue }
            let field = object["columnId"]?.stringValue ?? object["field"]?.stringValue ?? ""
            let direction = object["direction"]?.stringValue?.lowercased() ?? "asc"
            let a = resolveDerived(left, field) ?? .null
            let b = resolveDerived(right, field) ?? .null
            let comparison = derivedComparable(a).localizedStandardCompare(derivedComparable(b))
            if comparison != .orderedSame { return (comparison == .orderedAscending ? -1 : 1) * (direction == "desc" ? -1 : 1) }
        }
        return 0
    }
}

private enum DerivedDataSourceError: LocalizedError {
    case message(String)
    var errorDescription: String? { if case .message(let value) = self { return value }; return nil }
}

private func resolveDerived(_ row: [String: JSONValue], _ selector: String) -> JSONValue? {
    JSONValue(any: SelectorUtil.resolve(row.compactMapValues(\.anyValue), selector: selector))
}

private func derivedIdentity(_ row: [String: JSONValue], fields: [String]) -> String {
    fields.map { derivedComparable(resolveDerived(row, $0) ?? .null) }.joined(separator: "\u{001f}")
}

private func derivedNumber(_ value: JSONValue) -> Double? {
    switch value { case .number(let value): return value; case .string(let value): return Double(value); default: return nil }
}

private func derivedComparable(_ value: JSONValue) -> String {
    switch value {
    case .string(let value): return value
    case .number(let value): return String(format: "%020.8f", value)
    case .bool(let value): return value ? "1" : "0"
    case .null: return ""
    case .array(let values): return values.map(derivedComparable).joined(separator: ",")
    case .object(let values): return values.keys.sorted().map { "\($0)=\(derivedComparable(values[$0] ?? .null))" }.joined(separator: ",")
    }
}

private func deepMergeDerived(_ left: [String: JSONValue], _ right: [String: JSONValue]) -> [String: JSONValue] {
    var result = left
    for (key, value) in right { result[key] = value }
    return result
}

File: android/sdk/src/main/java/com/viant/forgeandroid/runtime/DerivedDataSourceRuntime.kt
package com.viant.forgeandroid.runtime

object DerivedDataSourceRuntime {
    fun run(sources: Map<String, List<Map<String, Any?>>>, spec: DerivedDataSourceSpec): List<Map<String, Any?>> {
        val limit = (spec.maxRows ?: 10_000).coerceAtLeast(1)
        spec.sources.forEach { ref -> require(sources[ref].orEmpty().size <= limit) { "Derived datasource source $ref exceeds maxRows $limit; use a server datasource." } }
        var rows = spec.sources.firstOrNull()?.let { sources[it] }.orEmpty().map { it.toMap() }
        spec.pipeline.forEach { step ->
            when (val operation = step.operation.lowercase()) {
                "filter" -> rows = rows.filter { evaluateDashboardCondition(step.whenCondition, form = it) }
                "select", "map" -> rows = rows.map { it + projectDerived(it, step) }
                "union" -> {
                    val ref = declaredSource(step.source, operation, spec)
                    rows = rows + sources[ref].orEmpty()
                }
                "join" -> rows = joinDerived(rows, sources[declaredSource(step.source, operation, spec)].orEmpty(), step, limit)
                "group" -> rows = groupDerived(rows, step)
                "sort" -> rows = rows.sortedWith { left, right -> compareDerived(left, right, step.orderBy) }
                else -> error("Unsupported derived operation: ${step.operation}")
            }
            require(rows.size <= limit) { "Derived datasource output exceeds maxRows $limit; use a server datasource." }
        }
        return rows
    }

    private fun declaredSource(raw: String?, operation: String, spec: DerivedDataSourceSpec): String {
        val ref = raw?.trim().orEmpty()
        require(ref.isNotEmpty() && ref in spec.sources) { "Derived $operation source ${ref.ifEmpty { "<empty>" }} must be declared in sources." }
        return ref
    }

    private fun joinDerived(left: List<Map<String, Any?>>, right: List<Map<String, Any?>>, step: DerivedDataStepSpec, limit: Int): List<Map<String, Any?>> {
        val fields = step.on.ifEmpty { listOf("id") }
        val rightByKey = linkedMapOf<String, MutableList<Map<String, Any?>>>()
        right.forEach { row ->
            val key = derivedIdentity(row, fields)
            require(step.joinCardinality.equals("many", true) || rightByKey[key].isNullOrEmpty()) { "Derived join expected one right row for key $key." }
            rightByKey.getOrPut(key) { mutableListOf() } += row
        }
        val result = mutableListOf<Map<String, Any?>>()
        left.forEach { row ->
            val matches = rightByKey[derivedIdentity(row, fields)].orEmpty()
            if (matches.isEmpty() && step.joinType.equals("inner", true)) return@forEach
            val selected = if (step.joinCardinality.equals("many", true)) matches.ifEmpty { listOf(emptyMap()) } else listOf(matches.firstOrNull().orEmpty())
            selected.forEach { match ->
                result += row + projectDerived(match, step)
                require(result.size <= limit) { "Derived datasource output exceeds maxRows $limit; use a server datasource." }
            }
        }
        return result
    }

    private fun groupDerived(rows: List<Map<String, Any?>>, step: DerivedDataStepSpec): List<Map<String, Any?>> =
        rows.groupBy { derivedIdentity(it, step.groupBy) }.toSortedMap().values.map { members ->
            val result = linkedMapOf<String, Any?>()
            step.groupBy.forEach { field -> result[field] = SelectorUtil.resolve(members.first(), field) }
            step.measures.forEach { measure ->
                val values = measure.source?.let { source -> members.mapNotNull { SelectorUtil.resolve(it, source) } }.orEmpty()
                result[measure.target] = when (measure.operation.lowercase()) {
                    "count" -> if (measure.source == null) members.size else values.size
                    "sum" -> values.sumOf { (it as? Number)?.toDouble() ?: it.toString().toDoubleOrNull() ?: 0.0 }
                    "min" -> values.minByOrNull(::derivedComparable)
                    "max" -> values.maxByOrNull(::derivedComparable)
                    "first" -> values.firstOrNull()
                    "list" -> values
                    else -> error("Unsupported derived measure: ${measure.operation}")
                }
            }
            result
        }

    private fun projectDerived(row: Map<String, Any?>, step: DerivedDataStepSpec): Map<String, Any?> {
        val result = linkedMapOf<String, Any?>()
        step.fields.forEach { (target, source) -> result[target] = if (source is kotlinx.serialization.json.JsonPrimitive && source.isString) SelectorUtil.resolve(row, source.content) else JsonUtil.elementToAny(source) }
        step.projections.forEach { projection -> result[projection.target] = projection.source?.let { SelectorUtil.resolve(row, it) } ?: projection.value?.let(JsonUtil::elementToAny) }
        return result
    }

    private fun compareDerived(left: Map<String, Any?>, right: Map<String, Any?>, orderBy: List<kotlinx.serialization.json.JsonElement>): Int {
        orderBy.forEach { raw ->
            val item = raw as? kotlinx.serialization.json.JsonObject ?: return@forEach
            val field = (item["columnId"] ?: item["field"])?.let(JsonUtil::elementToAny)?.toString().orEmpty()
            val direction = item["direction"]?.let(JsonUtil::elementToAny)?.toString()?.lowercase() ?: "asc"
            val value = derivedComparable(SelectorUtil.resolve(left, field)).compareTo(derivedComparable(SelectorUtil.resolve(right, field)))
            if (value != 0) return if (direction == "desc") -value else value
        }
        return 0
    }

    private fun derivedIdentity(row: Map<String, Any?>, fields: List<String>): String = fields.joinToString("\u001f") { derivedComparable(SelectorUtil.resolve(row, it)) }
    private fun derivedComparable(value: Any?): String = when (value) {
        null -> ""
        is Number -> String.format("%020.8f", value.toDouble())
        is Map<*, *> -> value.entries.sortedBy { it.key.toString() }.joinToString(",") { "${it.key}=${derivedComparable(it.value)}" }
        is Iterable<*> -> value.joinToString(",") { derivedComparable(it) }
        else -> value.toString()
    }
}
